api/saml/provider.py

```diff
@@ -0,0 +1,335 @@
+import datetime
+import json
+import logging
+
+from flask import url_for
+from flask_babel import lazy_gettext as _
+
+from api.authenticator import BaseSAMLAuthenticationProvider, PatronData
+from api.problem_details import *
+from api.saml.configuration import SAMLConfiguration, SAMLConfigurationSerializer, SAMLMetadataSerializer
+from api.saml.metadata import Subject, SubjectUIDExtractor, SubjectJSONEncoder
+from core.model import Credential, get_one_or_create, DataSource
+from core.util.problem_detail import ProblemDetail
+
+SAML_INVALID_SUBJECT = pd(
+    'http://librarysimplified.org/terms/problem/saml/invalid-subject',
+    status_code=401,
+    title=_('SAML invalid subject.'),
+    detail=_('SAML invalid subject.')
+)
+
+
+class SAMLAuthenticationProvider(BaseSAMLAuthenticationProvider):
```

Comment:
Are there going to be other subclasses of BaseSAMLAuthenticationProvider? I think you copied this design from OAuthAuthenticationProvider, which has one subclass -- for Clever -- but could conceivably have other subclasses for other systems that use OAuth. If any authentication provider that uses SAML can use SAMLAuthenticationProvider, you should be able to get rid of BaseSAMLAuthenticationProvider and just have one class.

Reply:
This is a very question. `SAMLAuthenticationProvider` implements Web Browser SSO Profile using two specific SAML bindings:
- HTTP-Redirect for SAML requests
- HTTP-Post for SAML responses

There may be other classes inherited from `BaseSAMLAuthenticationProvider` which could implement other SAML profiles or implement the same SSO profile but using other bindings (for example, SOAP).

Reply:
In that case, how about renaming `SAMLAuthenticationProvider` to `WebBrowserSAMLAuthenticationProvider` or something that conveys which part of SAML it implements?

Reply:
I renamed `SAMLAuthenticationProvider` to `SAMLWebSSOAuthenticationProvider`. Would you prefer to rename it to `WebBrowserSAMLAuthenticationProvider` instead?

Reply:
I must have missed that; SAMLWebSSOAuthenticationProvider is fine.